Server-errors-and-solutions.md

```diff
@@ -0,0 +1,80 @@
+* __Manual testing__: Having a functionality to generate dummy data for testing locally might help.
+
+## Corner Cases
+
+User-facing issues where certain corner case scenarios are not considered (empty list, undefined object etc.) Example: Deleting a hint that is active in the editor view caused console errors. Although these are not often serious, these issues can be tricky to debug due to lack of reproduction steps. The majority of server errors fall under this category. Other e.g. error #7907, #6298, #6199, #12528, #8702, #8464, #7970, #7546, #7444, #7727 (usage of plurals in certain languages), #7342, #7331, #7047, #6757, #6523 (didn't account for changes to interaction type), #5457
+
+### Preventive action
+
+* __Static analysis / compilation stage__: Strongly typed frontend will help catch some of these issues.
+* __Manual testing__: Developers generally test only the “happy” paths in their feature. We should encourage them to test edge cases. Adoption of template users in testing mentioned here should help uncover / test edge case scenarios.
+
+## Supporting old browsers
+Certain JS functionality may not be supported in older browsers. Example: String.prototype.match was used in the codebase but it lacked support in IE browsers. Fixes for these issues often involve adding a polyfill or swapping the relevant JS method with a more compatible one. Other e.g. errors #6909, #6174, #7926, #7794, #7784, #7205
+
+### Preventive action
+
+* __Static analysis / compilation stage__: Fix a JS standard that we expect all clients to support and specify a “target version” in typescript so that the code will be compiled to that version / JS standard.
+
+## Logic errors
+Logical errors in the code. Example: “this” scope was referring to the global scope when it was supposed to refer to a local one (#7873). These are quite rare. Other e.g. errors #7873, #7872 (incorrect scope of this), #7927, #7460
+
+### Preventive action
+
+* __Code review__: These issues need to be flagged during code reviews. Often happens when the developer is refactoring existing code but hasn’t fully tested their changes.
+
+## Data-related Issue
+
+Issues that resulted in corrupt or missing data on the server. Issues like this are quite hard to debug and involve complicated one-off jobs to be run. Example: Certain exploration stats models were missing on the production server, this resulted in few deferred tasks erroring often and piling up in the stackdriver logs. Fortunately, these issues are quite rare. Other e.g. errors #6159, #9632, #7440
+
+### Preventive action
+
+* __Manual testing__: When testing, perform destructive operations on entities that have related supplementary entities to check if the supplementary objects reflect the correct state of the original ones.
+Testing with foreign / special characters (as mentioned under Template Users).
```

Comment:
You should have 1., 2.  numbered bullet points here, similar to the original doc. Please preview your changes to make sure they are correct.

Reply:
It didn't have any heading so I thought it will be another new line in manual testing para itself. 
okay Now I changed it